app/services/PageRenderer.scala

```diff
@@ -31,26 +31,41 @@ class PageRenderer @Inject() () {
   }
 
   def renderPagePostSubmit(page: Page, labels: Labels, answer: String): (Option[String], Labels) = {
-    implicit val stanzaMap: Map[String, Stanza] = page.keyedStanzas.map(ks => (ks.key, ks.stanza)).toMap
 
     @tailrec
-    def evaluatePostInputStanzas(next: String, labels: Labels, seen: Seq[String]): (Option[String], Labels) =
+    def evaluatePostInputStanzas(next: String, labels: Labels, seen: Seq[String], stanzaMap: Map[String, Stanza]): (Option[String], Labels) = {
       if (next == page.id) (None, labels)                     // next indicates current page
       else stanzaMap.get(next) match {
         case None => (Some(next), labels)
         case Some(_) if seen.contains(next) => (None, labels) // Legacy: Interpret redirect to stanza prior to input as current page
         case Some(s) => s match {
-          case EndStanza => (Some(next), labels)
+          case EndStanza =>
+            labels.takeFlow match {
+              case Some((nxt, stanzas, updatedLabels)) =>
+                evaluatePostInputStanzas(nxt, updatedLabels, seen, stanzaMap ++ stanzas.map(ks => (ks.key, ks.stanza)).toMap)
+              case None => (Some(next), labels)
+            }
           case s: Stanza with Evaluate =>
             val (next, updatedLabels) = s.eval(labels)
-            evaluatePostInputStanzas(next, updatedLabels, seen)
+            evaluatePostInputStanzas(next, updatedLabels, seen, stanzaMap)
         }
       }
+    }
+
+    val stanzaMap: Map[String, Stanza] = page.keyedStanzas.map(ks => (ks.key, ks.stanza)).toMap
+    val (_, newLabels, seen, nextPageId, optionalInput) = evaluateStanzas(stanzaMap(page.id).next.head, labels, Nil, Nil)(stanzaMap)
```

Comment:
Do you need the two Nils here?

Reply:
could be removed yes and rely on the default args, will update